lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py

```diff
@@ -0,0 +1,409 @@
+                    value = bitmath.Byte(float(memory[:-1]) / 1000).to_GiB().value
+                else:
+                    if not memory.endswith("B"):
+                        # Bitmath doesn't recognize e.g. Gi/G - needs to be GiB/GB.
+                        memory += "B"
+                    value = bitmath.parse_string(memory).to_GiB().value
+
+                if value < 0:
+                    raise ValueError(
+                        "Failed to apply memory resource limit patch: "
+                        "value must be greater than or equal to 0"
+                    )
+                copy["memory"] = f"{str(value)}Gi"
+
+            if cpu := copy.get("cpu"):
+                # TODO need to take into account m, k/K, M, G, T, P and E, but for CPU count it's
+                # probably ok to only support "m" and plain decimals for now.
+                # https://github.com/gtsystem/lightkube/issues/36
+                if cpu.endswith("m"):
+                    # This is milli. Divide by 1000.
+                    value = int(cpu[:-1]) / 1000.0
+                else:
+                    # Round up to whole millis (e.g. 0.30000000000000004 -> 0.301)
+                    value = ceil(float(cpu) * 1000) / 1000.0
+
+                if value < 0:
+                    raise ValueError(
+                        "Failed to apply cpu resource limit patch: "
+                        "value must be greater than or equal to 0"
+                    )
+                copy["cpu"] = str(value)
+
+            return copy
+
+        return ResourceRequirements(limits=_conv(res_req.limits), requests=_conv(res_req.requests))  # type: ignore[arg-type]
+
+    def is_ready(self):
+        """Reports if the resource patch has been applied and is in effect.
+
+        Returns:
+            bool: A boolean indicating if the service patch has been applied and is in effect.
+        """
+        client = Client()
+        pod = client.get(Pod, name=self._pod, namespace=self._namespace)
+        podspec = self._get_container(self._container_name, pod.spec.containers)  # type: ignore[attr-defined]
+
+        try:
+            ready = self._conv_res_req(self.resource_reqs) == self._conv_res_req(podspec.resources)
+            patched = self._is_patched(client)
+        except (ValueError, ApiError) as e:
+            msg = f"Failed to apply resource limit patch: {e}"
+            logger.error(msg)
+            self.on.patch_failed.emit(message=msg)
+            return False
+
+        return patched and ready
+
+    @classmethod
+    def _get_container(cls, container_name: str, containers: List[Container]) -> Container:
+        """Find our container from the container list, assuming list is unique by name.
+
+        Typically, *.spec.containers[0] is the charm container, and [1] is the (only) workload.
+
+        Raises:
+            StopIteration, if the user-provided container name does not exist in the list.
```

Comment:
Perhaps a more appropriate exception should be raised?

Reply:
ValueError or a custom ContainerNotFoundError

Reply:
> ContainerNotFoundError

I prefer this one

Reply:
I feel that ContainerNotFound *would* likely be more helpful when encountered in the logs. It would immediately tell you what the issue is (or at least help narrow it down).

Reply:
Changed to custom exception.